pkg/controllers/management/globaldns/globaldns_handler.go

```diff
@@ -0,0 +1,228 @@
+	ingressLister     v1beta1Rancher.IngressLister
+	ctx               context.Context
+	managementContext *config.ManagementContext
+}
+
+func newGlobalDNSController(ctx context.Context, mgmt *config.ManagementContext) *GlobalDNSController {
+	extInterface, err := v1beta1Rancher.NewForConfig(mgmt.RESTConfig)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	n := &GlobalDNSController{
+		globalDNSInterface: mgmt.Management.GlobalDNSs(""),
+		globalDNSLister:    mgmt.Management.GlobalDNSs("").Controller().Lister(),
+		ctx:                ctx,
+		ingressLister:      extInterface.Ingresses("default").Controller().Lister(),
+		managementContext:  mgmt,
+	}
+	return n
+}
+
+func (n *GlobalDNSController) getIngressInterface(namespace string) clientv1beta1.IngressInterface {
+	if namespace != "" {
+		return n.managementContext.K8sClient.Extensions().Ingresses(namespace)
+	}
+	return nil
+}
+
+func Register(ctx context.Context, management *config.ManagementContext) {
+	n := newGlobalDNSController(ctx, management)
+	management.Management.GlobalDNSs("").AddHandler(ctx, GlobaldnsController, n.sync)
+}
+
+//sync is called periodically and on real updates
+func (n *GlobalDNSController) sync(key string, obj *v3.GlobalDNS) (runtime.Object, error) {
+	logrus.Debugf("GlobalDNSController called")
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	//check if status.endpoints is set, if yes create a dummy ingress if not already present
+	//if ingress exists, update endpoints if different
+
+	if len(obj.Status.Endpoints) > 0 {
+		ingress, err := n.getIngressForGlobalDNS(obj)
+		if err != nil {
+			logrus.Errorf("GlobalDNSController: Error listing ingress for the GlobalDNS %v", err)
+			return nil, err
+		}
+		if ingress == nil {
+			ingress, err = n.createIngressForGlobalDNS(obj)
+			if err != nil {
+				logrus.Errorf("GlobalDNSController: Error creating an ingress for the GlobalDNS %v", err)
+				return nil, err
+			}
+		}
+		if ingress != nil {
+			err = n.updateIngressEndpoints(ingress, obj.Status.Endpoints)
+			if err != nil {
+				logrus.Errorf("GlobalDNSController: Error updating ingress for the GlobalDNS %v", err)
+				return nil, err
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (n *GlobalDNSController) getIngressForGlobalDNS(globaldns *v3.GlobalDNS) (*v1beta1.Ingress, error) {
+	ingresses, err := n.ingressLister.List(globaldns.Namespace, labels.NewSelector())
+	if err != nil {
+		return nil, err
+	}
+	logrus.Infof("listed ingress %v for globalDNS ns %s", len(ingresses), globaldns.Namespace)
+
+	for _, ingress := range ingresses {
+		//mark the service which own by ingress but not related to ingress
+		if n.isIngressOwnedByGlobalDNS(ingress, globaldns) {
+			return ingress, nil
+		}
+	}
+	return nil, nil
+}
+
+func (n *GlobalDNSController) isIngressOwnedByGlobalDNS(ingress *v1beta1.Ingress, globaldns *v3.GlobalDNS) bool {
+	for i, owners := 0, ingress.GetOwnerReferences(); owners != nil && i < len(owners); i++ {
```

Comment:
I think we can apply this check on top of `Can we perhaps name the ingress with global dns name then? And use get for fetching it`

Reply:
done